viewer/src/components/ResearchTile.vue

```diff
@@ -50,31 +50,48 @@
 
 <script lang="ts">
 import Vue from "vue";
-import { Component, Prop } from "vue-property-decorator";
-import {
+import {Component, Prop} from "vue-property-decorator";
+import Engine, {
+  AdvTechTilePos,
+  canResearchField,
+  canTakeAdvancedTechTile,
   Condition,
   Event,
   Federation,
   Operator,
   Planet as PlanetEnum,
   Player,
+  PlayerEnum,
   ResearchField,
   researchTracks,
   Resource as ResourceEnum,
   Reward,
 } from "@gaia-project/engine";
-import { descriptions } from "../data/research";
+import {descriptions} from "../data/research";
 import Token from "./Token.vue";
 import FederationTile from "./FederationTile.vue";
 import Planet from "./Planet.vue";
 import Resource from "./Resource.vue";
 
+function tokenClass(engine: Engine, player: Player, level: number, field: ResearchField): string {
+  if (level < 4) {
+    return "";
+  }
+  const tilePos = "adv-" + field as AdvTechTilePos;
+  if (canTakeAdvancedTechTile(engine, player.data, tilePos) || canResearchField(engine, player, field)) {
+    return "warn";
+  }
+  return "";
+}
+
 @Component<ResearchTile>({
   computed: {
-    players(): Player[] {
-      return this.$store.state.gaiaViewer.data.players.filter(
-        (player) => player.faction && player.data.research[this.field] === this.level
-      );
+    players(): Array<{ player: Player; class: string }> {
+      const data = this.$store.state.gaiaViewer.data;
+      const players = data.players;
+      return players
+        .filter((player) => player.faction && player.data.research[this.field] === this.level)
+        .map((p) => ({player: p, class: tokenClass(data, p, this.level, this.field)}));
     },
```

Comment:
This is old code. The functoins in `computed` can actually move to getters, it would be cleaner and keep the type information. And you wouldn't need to move tokenClass outside of the class